File: typescript/behavioral_pattern/src/template_method/Beverage.ts
abstract class Beverage {
    public _prepareBeverage(): void {
        this.boilWater();
        this.brew();
        this.pourInCup();
        if (this.customerWantsCondiments()) {
            this.addCondiments();
        }
        this.serve();
    }

    private boilWater(): void {
        console.log("Boiling water...");
    }

    private pourInCup(): void {
        console.log("Pouring into cup...");
    }

    private serve(): void {
        console.log("Your beverage is ready!");
    }

    protected abstract brew(): void;
    protected abstract addCondiments(): void;

    protected customerWantsCondiments(): boolean {
        return true;
    }
}

export default Beverage;

File: behavioral_pattern/template_method/Beverage.java
package template_method;

public abstract class Beverage {
    public final void prepareBeverage() {
        boilWater();
        brew();
        pourInCup();
        if (customerWantsCondiments()) {
            addCondiments();
        }
        serve();
    }

    private void boilWater() {
        System.out.println("Boiling water...");
    }

    private void pourInCup() {
        System.out.println("Pouring into cup...");
    }

    private void serve() {
        System.out.println("Your beverage is ready!");
    }

    protected abstract void brew();
    protected abstract void addCondiments();

    protected boolean customerWantsCondiments() {
        return true;
    }
}